File: derp/Lexer.py
import re

class Lexer:

  class Token:
    def __init__( self, name, value ):
      self.name = name
      self.value = value
      
    def __eq__( self, token ):
      return self.name == token.name and self.value == token.value
      
    def __repr__( self ):
      return 'Token( %s, \'%s\' )' % ( self.name, self.value )
      
  class Rule:
    def __init__( self, name, pattern, exclude ):
      self.name = name
      self.pattern = pattern
      self.exclude = exclude
      
    def __repr__( self ):
      return 'Rule( %s, %s, %s )' % ( self.name, self.pattern, self.exclude )
      
    def match( self, input ):
      result = re.match( self.pattern, input )
      if result:
        return Lexer.Token( self.name, result.group() )

  def __init__( self, rules ):
    self.rules = rules
    
  def lex( self, input ):
    pos = 0
    toks = []
    while pos < len( input ):
      rawResults = [ ( rule, rule.match( input[pos:] ) ) for rule in self.rules ]
      results = filter( lambda result: result[1], rawResults )
      assert results, 'no results in %s for %s' % ( rawResults, input[pos:] )
      bestResult = max( results, key = lambda result: len( result[1].value ) )
      if not bestResult[0].exclude:
        toks.append( bestResult[1] )
      pos += len( bestResult[1].value )
    return toks

if __name__ == '__main__':
  lexer = Lexer( [ Lexer.Rule( 'lparen', '\(', False ),
                   Lexer.Rule( 'rparen', '\)', False ),
                   Lexer.Rule( 'int', '-?\d+', False ), 
                   Lexer.Rule( 'float', '-?\d+\.\d+', False ),
                   Lexer.Rule( 'id', '\w+', False ),
                   Lexer.Rule( 'ws', '\s+', True ),
                 ] )
  results = lexer.lex( '( foo 1 -3.14 )' )
  assert results == [ Lexer.Token( 'lparen', '(' ),
                      Lexer.Token( 'id', 'foo' ), 
                      Lexer.Token( 'int', '1' ), 
                      Lexer.Token( 'float', '-3.14' ),
                      Lexer.Token( 'rparen', ')' ),
                    ]

File: trunk/derp2/derp/Lexer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Text.RegularExpressions;

namespace derp
{
    public class Lexer
    {
        public class Result
        {
            public Rule Rule { get; set; }

            public string Value { get; set; }

            public override string ToString()
            {
                return "Lexer.Result( " + Rule.Name + ", " + Value + " )";
            }
        }

        public class Rule
        {
            public string Name { get; set; }

            public string Pattern { get; set; }

            public bool Include { get; set; }

            private Regex regex = null;

            public Rule(string name, string pattern, bool include)
            {
                Name = name;
                Pattern = pattern;
                Include = include;
            }

            public Result Apply(string input, int pos)
            {
                if (regex == null)
                {
                    regex = new Regex(Pattern);
                }
                Match match = regex.Match(input, pos);
                if (match.Success && match.Index == pos && match.Length > 0)
                {
                    return new Result() { Rule = this, Value = input.Substring(pos, match.Length) };
                }
                else
                {
                    return null;
                }
            }

            public override string ToString()
            {
                return "Lexer.Rule( " + Name + ", " + Pattern + ", " + Include + " )";
            }
        }

        private List<Rule> rules = new List<Rule>();

        public List<Rule> Rules
        {
            get { return rules; }
            set { rules = value; }
        }

        public Lexer(params Rule[] rules)
        {
            Rules = rules.ToList();
        }

        public List<Result> Lex(string input)
        {
            List<Result> results = new List<Result>();
            for (int pos = 0; pos < input.Length; )
            {
                Result result = rules
                    .Select(rule => rule.Apply(input, pos))
                    .Where( r => r != null )
                    .FirstOrDefault();
                if (result == null)
                {
                    throw new Exception("lex error " + input.Substring(pos));
                }
                else
                {
                    if (result.Rule.Include)
                    {
                        results.Add(result);
                    }
                    pos += result.Value.Length;
                }
            }
            return results;
        }
    }
}
